contrib/bloop/src/mill/contrib/bloop/BloopImpl.scala

```diff
@@ -188,6 +188,27 @@ class BloopImpl(ev: () => Evaluator, wd: os.Path) extends ExternalModule { outer
         case None => T.task(None)
       }).map(_.getOrElse(Config.LinkerMode.Debug))
 
+    // //////////////////////////////////////////////////////////////////////////
+    //  Classpath
+    // //////////////////////////////////////////////////////////////////////////
+
+    val ivyDepsClasspath = module.resolvedIvyDeps.map(_.map(_.path).toSeq)
+    val runIvyDepsClasspath = module.resolvedRunIvyDeps.map(_.map(_.path).toSeq)
+
+    def transitiveClasspath(m: JavaModule): Task[Seq[os.Path]] = T.task {
+      (m.moduleDeps ++ m.compileModuleDeps).map(classes) ++
+        m.unmanagedClasspath().map(_.path) ++
+        T.traverse(m.moduleDeps ++ m.compileModuleDeps)(transitiveClasspath)().flatten
```

Comment:
I don't think adding `compileModuleDeps` transitively is the right thing here. Compile-time dependencies should only be relevant at compilation time and typically don't be part of the transitive dependencies.

Reply:
> I have two concerns:
> 
> First, I think the transitive dependency calculation isn't correct, because it includes also compile-time dependencies transitively, which is wrong. I also realize this issue was already present before your PR, but as you touched this code, I wanted to point it out. And at this point, I'd like to request a test case resembling a project with multiple modules and ensuring the classpaths and run-classpaths are correct, which brings me to the other point.
> 
> This is a contrib module, where we don't enforce the same strict quality standards. I guess this means, we can just merge as is, with missing tests and such. But as Bloop is used by many users (I guess), we should address my first concern.

Thanks for getting back on the issue! Yeah, I was a bit suspicious of the quality of the class as well, but to be honest I don't know exactly what needs to go where.

What I do know is that the classpath of the jvm-specific bloop configuration needs to contain all compiled classes that the module depends on, as well as any runtime dependencies. So it's basically a superset of the project classpath if I understand things correctly. 

Are you saying it is correct that the classpath consist of module.ivyDeps ++ compiled moduleDeps classes but without their own corresponding classpath dependencies (and ++ module.runIvyDeps for the jvm-part of course)?

Reply:
If we trust the SBT implementation I could try to create a multi-project SBT configuration and have a look at the generated bloop-files from that, and try to align the output of the BloopImpl.

Reply:
I can't say anything about the sbt implementation.

Compile-time dependencies should only be visible at compile time. Once compilation is done and the compiled classes are part of the classpath, compile-time dependencies should no longer be present. If you need the same dependency (or an alternative) at runtime, you need to add them to runIvyDeps.

There is already a test suite for Bloop (`mill.contrib.bloop.BloopTests`), so best would be to review or add to it, so that we can assert the wanted outcome. If it helps to cross-check against SBT, I'd say, go for it. Ideally, with the exact same test project. The more consistent our ecosystem is, the better.

Reply:
> Are you saying it is correct that the classpath consist of module.ivyDeps ++ compiled moduleDeps classes but without their own corresponding classpath dependencies (and ++ module.runIvyDeps for the jvm-part of course)?

This is correct for compile-time! At runtime, we should leave the `compileModuleDeps` out.